### PR TITLE
[X86] Handle shifts + and in `LowerSELECTWithCmpZero`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[X86] Add tests support shifts + and in `LowerSELECTWithCmpZero`; NFC**
- **[X86] Handle shifts + and in `LowerSELECTWithCmpZero`**

shifts are the same as sub where rhs == 0 is identity.
and is the inverted case where:
    `SELECT (AND(X,1) == 0), (AND Y, Z), Y`
        -> `(AND Y, (OR NEG(AND(X, 1)), Z))`
With -1 as the identity.

### CODE REVIEW

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by RKSimon** [comment]
Thanks @goldsteinn - I'm trying to find profitable uses for these folds on CMOV targets - any ideas?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -24136,15 +24141,40 @@ static SDValue LowerSELECTWithCmpZero(SDValue CmpVal, SDValue LHS, SDValue RHS,
...
-      SDValue Mask = SplatLSB();
-      SDValue And = DAG.getNode(ISD::AND, DL, VT, Mask, Src1); // Mask & z
-      return DAG.getNode(RHS.getOpcode(), DL, VT, Src2, And);  // y Op And
+    // SELECT (AND(X,1) == 0), Y, (SHL Y, Z) -> (SHL Y, (AND NEG(AND(X,1)), Z))
+    // SELECT (AND(X,1) == 0), Y, (SRA Y, Z) -> (SRA Y, (AND NEG(AND(X,1)), Z))
+    // SELECT (AND(X,1) == 0), Y, (SRL Y, Z) -> (SRL Y, (AND NEG(AND(X,1)), Z))
+    if (!Subtarget.canUseCMOV() && isIdentityPatternZero()) {
+      SDValue Mask = SplatLSB(Src1.getValueType());
+      SDValue And = DAG.getNode(ISD::AND, DL, Src1.getValueType(), Mask,
+                                Src1); // Mask & z
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:24170]: @goldsteinn clang-format this

**Comment by goldsteinn** [comment]
> Thanks @goldsteinn - I'm trying to find profitable uses for these folds on CMOV targets - any ideas?

I think probably most of the profitable cases are already handled in the middle end i.e `foldSelectICmpAndBinOp` or `foldSelectICmpEq`.

If we are generating these in the backend then the cases we handle in the middle end are probably roughly what is profitable even with CMOV.

**Pull request closed without merging**